src/particles.zig

```diff
@@ -0,0 +1,385 @@
+		shader = Shader.initAndGetUniforms("assets/cubyz/shaders/particles/particles.vs", "assets/cubyz/shaders/particles/particles.fs", "", &uniforms);
+
+		properties = EmmiterProperties{
+			.gravity = .{0, 0, 4},
+			.drag = 2,
+			.lifeTimeMin = 5,
+			.lifeTimeMax = 5,
+			.velMin = 5,
+			.velMax = 10,
+		};
+		particles = main.globalAllocator.alloc(Particle, maxCapacity);
+		particlesSSBO = SSBO.init();
+		particlesSSBO.createDynamicBuffer(maxCapacity*@sizeOf(Particle));
+		particlesSSBO.bind(13);
+
+		seed = @bitCast(@as(i64, @truncate(std.time.nanoTimestamp())));
+	}
+
+	pub fn deinit() void {
+		shader.deinit();
+		particlesSSBO.deinit();
+		particles.len = maxCapacity;
+		main.globalAllocator.free(particles);
+	}
+
+	pub fn update(deltaTime: f32) void {
+		const vdt: Vec3f = @as(Vec3f, @splat(deltaTime));
+
+		var i: u32 = 0;
+		while(i < particles.len) {
+			var particle = particles[i];
+			particle.lifeLeft -= deltaTime;
+			if(particle.lifeLeft < 0) {
+				particles[i] = particles[particles.len - 1];
+				particles.len -= 1;
+				continue;
+			}
+
+			particle.vel += properties.gravity*vdt;
+			particle.vel *= @splat(@max(0, 1 - properties.drag*deltaTime));
+			const vel = particle.vel*vdt;
+
+			// TODO: OPTIMIZE THE HELL OUT OF THIS
+			if(particle.collides) {
+				const hitBox: game.collision.Box = .{.min = @splat(particle.size*-0.5), .max = @splat(particle.size*0.5)};
+				particle.pos[0] += vel[0];
+				if(game.collision.collides(.client, .x, -vel[0], particle.pos, hitBox)) |box| {
+					if(vel[0] < 0) {
+						particle.pos[0] = @floatCast(box.max[0] - hitBox.min[0]);
+					} else {
+						particle.pos[0] = @floatCast(box.min[0] - hitBox.max[0]);
+					}
+				}
+				particle.pos[1] += vel[1];
+				if(game.collision.collides(.client, .y, -vel[1], particle.pos, hitBox)) |box| {
+					if(vel[1] < 0) {
+						particle.pos[1] = @floatCast(box.max[1] - hitBox.min[1]);
+					} else {
+						particle.pos[1] = @floatCast(box.min[1] - hitBox.max[1]);
+					}
+				}
+				particle.pos[2] += vel[2];
+				if(game.collision.collides(.client, .z, -vel[2], particle.pos, hitBox)) |box| {
+					if(vel[2] < 0) {
+						particle.pos[2] = @floatCast(box.max[2] - hitBox.min[2]);
+					} else {
+						particle.pos[2] = @floatCast(box.min[2] - hitBox.max[2]);
+					}
+				}
+			} else {
+				particle.pos += vel;
+			}
+
+			// TODO: optimize
+			const intPos: Vec3i = @intFromFloat(@floor(particle.pos));
+			const light: [6]u8 = main.renderer.mesh_storage.getLight(intPos[0], intPos[1], intPos[2]) orelse @splat(0);
+			particle.light = (@as(u32, light[0] >> 3) << 25 |
+				@as(u32, light[1] >> 3) << 20 |
+				@as(u32, light[2] >> 3) << 15 |
+				@as(u32, light[3] >> 3) << 10 |
+				@as(u32, light[4] >> 3) << 5 |
+				@as(u32, light[5] >> 3));
+
+			particles[i] = particle;
+			i += 1;
+		}
+	}
+
+	pub fn spawn(id: []const u8, count: u32, pos: Vec3f, size: f32, collides: bool, shape: EmmiterShape) void {
```

Comment:
Could you optimize all Emitter related parameters into a separate emitter struct?
Also please move the bulk of this function there, implementation details for all the shapes don't belong into the ParticleSystem struct in my opinion.